test/unit/BPool/BPool_ExitPool.t.sol

```diff
@@ -0,0 +1,135 @@
+    // it should revert
+    //     division by zero
+    vm.expectRevert(BNum.BNum_DivZero.selector);
+    bPool.exitPool(0, minAmountsOut);
+  }
+
+  function test_RevertWhen_PoolAmountInIsTooSmall(uint256 amountIn) external {
+    amountIn = bound(amountIn, 0, (INIT_POOL_SUPPLY / 1e18) / 2 - 1);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidPoolRatio.selector);
+    bPool.exitPool(amountIn, minAmountsOut);
+  }
+
+  function test_RevertWhen_CallerDoesNotHaveTheRequiredPoolShares() external {
+    // it should revert
+    vm.expectRevert(
+      abi.encodeWithSelector(IERC20Errors.ERC20InsufficientBalance.selector, address(this), 0, poolAmountIn)
+    );
+    bPool.exitPool(poolAmountIn, minAmountsOut);
+  }
+
+  modifier whenCallerHasEnoughPoolShares() {
+    // setup leaves them in bpool contract, this moves them to the caller
+    bPool.call__pushPoolShare(address(this), INIT_POOL_SUPPLY);
+    _;
+  }
+
+  function test_RevertWhen_BalanceOfPoolInAnyTokenIsZero() external whenCallerHasEnoughPoolShares {
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(0)));
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidTokenAmountOut.selector);
+    bPool.exitPool(poolAmountIn, minAmountsOut);
+  }
+
+  function test_RevertWhen_ReturnedAmountOfATokenIsLessThanMinAmountsOut() external whenCallerHasEnoughPoolShares {
+    minAmountsOut[1] += 1;
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenAmountOutBelowMinAmountOut.selector);
+    bPool.exitPool(poolAmountIn, minAmountsOut);
+  }
+
+  function test_WhenPreconditionsAreMet() external whenCallerHasEnoughPoolShares {
+    // it pulls poolAmountIn shares
+    bPool.expectCall__pullPoolShare(address(this), poolAmountIn);
+    // it sends zero exitFee to factory
+    // NOTE: if exit_fee is going to be hard-coded to zero we could remove this
+    // and save some gas
+    bPool.expectCall__pushPoolShare(deployer, 0);
+    // it burns pool shares
+    bPool.expectCall__burnPoolShare(poolAmountIn);
+    // it calls _pushUnderlying for every token
+    bPool.mock_call__pushUnderlying(token, address(this), expectedToken0Out);
+    bPool.expectCall__pushUnderlying(token, address(this), expectedToken0Out);
+    bPool.mock_call__pushUnderlying(secondToken, address(this), expectedToken1Out);
+    bPool.expectCall__pushUnderlying(secondToken, address(this), expectedToken1Out);
+    // it sets the reentrancy lock
+    bPool.expectCall__setLock(_MUTEX_TAKEN);
+    // // it emits LOG_CALL event
+    bytes memory _data = abi.encodeWithSelector(IBPool.exitPool.selector, poolAmountIn, minAmountsOut);
+    vm.expectEmit();
+    emit IBPool.LOG_CALL(IBPool.exitPool.selector, address(this), _data);
+    // it emits LOG_EXIT event for every token
+    vm.expectEmit();
+    emit IBPool.LOG_EXIT(address(this), token, expectedToken0Out);
+    vm.expectEmit();
+    emit IBPool.LOG_EXIT(address(this), secondToken, expectedToken1Out);
```

Comment:
how do you feel about packing all `for every token` in a common loop? (sometimes it'll need to happen before and after the call)

Reply:
for two tokens I don't think the required setup (putting amounts into an array) is more readable or concise than just inlining it.

having everything in a single loop could make it more _parameterizable_ (running the test with different data by choosing a different array) but I don't think the exitpool tests would benefit from it, unlike, say, BMath